rest/route/pod_agent.go

```diff
@@ -328,3 +348,155 @@ func (h *podAgentNextTask) findDispatcher() (*dispatcher.PodDispatcher, error) {
 
 	return pd, nil
 }
+
+// POST /rest/v2/pods/{pod_id}/task/{task_id}/end
+
+type podAgentEndTask struct {
+	env     evergreen.Environment
+	podID   string
+	taskID  string
+	details apimodels.TaskEndDetail
+}
+
+func makePodAgentEndTask(env evergreen.Environment) gimlet.RouteHandler {
+	return &podAgentEndTask{
+		env: env,
+	}
+}
+
+func (h *podAgentEndTask) Factory() gimlet.RouteHandler {
+	return &podAgentEndTask{
+		env: h.env,
+	}
+}
+
+func (h *podAgentEndTask) Parse(ctx context.Context, r *http.Request) error {
+	h.podID = gimlet.GetVars(r)["pod_id"]
+	if h.podID == "" {
+		return errors.New("missing pod ID")
+	}
+	h.taskID = gimlet.GetVars(r)["task_id"]
+	if h.taskID == "" {
+		return errors.New("missing task ID")
+	}
+	details := apimodels.TaskEndDetail{}
+	if err := utility.ReadJSON(utility.NewRequestReader(r), &details); err != nil {
+		return errors.Wrap(err, "reading task end details from JSON request body")
+	}
+	h.details = details
+
+	// Check that status is either finished or aborted (i.e. undispatched)
+	if !evergreen.IsValidTaskEndStatus(details.Status) && details.Status != evergreen.TaskUndispatched {
+		return errors.Errorf("invalid end status '%s' for task %s", details.Status, h.taskID)
+	}
+	return nil
+}
+
+// Run finishes the given task and generates a job to collect task end stats.
+// It then marks the task as finished or inactive if aborted.
+func (h *podAgentEndTask) Run(ctx context.Context) gimlet.Responder {
+	finishTime := time.Now()
+	p, err := findPod(h.podID)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
```

Comment:
`gimlet.MakeJSONInternalErrorResponder` - if DB find methods return non-nil error, it would be due to an error other than the doc-not-found error, which is considered a server-side issue. `gimlet.MakeJSONErrorResponder` will return 400 Bad Request, which is supposed to be used for client-side errors, so using `gimlet.MakeJSONInternalErrorResponder` is generally preferred since it indicates a generic server-side error.